### PR TITLE
[BEAM-2859] Fixed processing timers not being properly fired in spark-runner

### PR DESCRIPTION
**Pull request by staslev**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by staslev** [comment]
Run Spark ValidatesRunner

**Comment by staslev** [comment]
The following changes were performed as part of this PR:

1. `TimeDomain.PROCESSING_TIME` based timers are always passed in to `ReduceFnRunner` regardless of the watermark position.
2. Timers are no longer deleted from `SparkTimerInternals` by (what used to be) `SparkTimerInternals#getTimersReadyToProcess()`, they are deleted by GC timer.
3. `SparkTimerInternals#getTimersReadyToProcess()` was renamed to `filterEligibleForProcessing()` and moved to `SparkGroupAlsoByWindowViaWindowSet` since it is the one who determines what *eligible* actually means.

**Comment by staslev** [comment]
Run Spark ValidatesRunner

**Comment by staslev** [comment]
Retest this please

**Comment by staslev** [comment]
Retest this please

**Comment by staslev** [comment]
Run Spark ValidatesRunner

**Comment by staslev** [comment]
Run Spark ValidatesRunner

**Comment by staslev** [comment]
R: @amitsela 
CC: @kennknowles

File: runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java [hunk 1]
```diff
@@ -204,6 +207,34 @@ public void outputWindowedValue(
         this.droppedDueToLateness = droppedDueToLateness;
       }
 
+      /**
+       * Retrieves the timers that are eligible for processing by {@link
+       * org.apache.beam.runners.core.ReduceFnRunner}. Even if a given timer is deemed eligible for
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java]: I'm not sure if the comment on firing is in place here.
I would add the comment you put in `@return` and leave only the first sentence there: `A collection of timers...`

Comment (staslev) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java]: 👍 
I'll move the part in questions to the call site of `reduceFnRunner.onTimers(...)`.

**Comment by staslev** [comment]
Retest this please

**Pull request closed without merging**